drpg/sync.py

```diff
@@ -56,64 +59,119 @@ def sync(self) -> None:
         self._api.token()
         logger.info("Fetching products list")
-        process_item_args = (
-            (product, item)
-            for product in self._api.customer_products()
-            for item in product["files"]
-            if self._need_download(product, item)
+
+        q: queue.PriorityQueue = queue.PriorityQueue()
+
+        for product in self._api.products(1):
+            q.put(QueueItem(QueueItemType.EXTRACT_ITEMS, (product,)))
+
+        pool = ThreadPool(self._config.threads)
+        q.put(QueueItem(QueueItemType.GET_PAGE, (2,)))
+
+        pool.starmap_async(
+            self._process,
+            ((q,) for _ in range(self._config.threads)),
+            error_callback=print,  # TODO Add a real errback
         )
 
-        with ThreadPool(self._config.threads) as pool:
-            pool.starmap(self._process_item, process_item_args)
+        q.join()
+        pool.terminate()
+
         logger.info("Done!")
 
-    @suppress_errors(httpx.HTTPError, PermissionError)
-    def _process_item(self, product: Product, item: DownloadItem) -> None:
+    def _process(self, q: queue.PriorityQueue) -> None:  # noqa: C901 XXX
+        while True:
+            try:
+                queue_item: QueueItem = q.get(block=True, timeout=0.5)
+            except queue.Empty:
+                continue
+
+            if queue_item.action == QueueItemType.GET_PAGE:
+                page, *_ = queue_item.args
+                products = self._api.products(page)
+                try:
+                    product = next(products)
+                except StopIteration:
+                    pass
+                else:
+                    q.put(QueueItem(QueueItemType.EXTRACT_ITEMS, (product,)))
+                    q.put(QueueItem(QueueItemType.GET_PAGE, (page + 1,)))
+
+                for product in products:
+                    q.put(QueueItem(QueueItemType.EXTRACT_ITEMS, (product,)))
+
+            elif queue_item.action == QueueItemType.EXTRACT_ITEMS:
+                product, *_ = queue_item.args
+                for item in product["files"]:
+                    q.put(QueueItem(QueueItemType.PREPARE_DOWNLOAD_URL, (product, item)))
+            elif queue_item.action == QueueItemType.PREPARE_DOWNLOAD_URL:
+                product, item, *_ = queue_item.args
+                ready, url = self._prepare_download_url(product, item)
+                if ready and url is not None:
+                    path = self._file_path(product, item)
+                    q.put(QueueItem(QueueItemType.DOWNLOAD, (url, path)))
+            elif queue_item.action == QueueItemType.DOWNLOAD:
+                url_data, path, *_ = queue_item.args
+                self._download_from_url(url_data, path)
+            q.task_done()
+
+    #  @suppress_errors(httpx.HTTPError, PermissionError)
+    def _prepare_download_url(
+        self, product: Product, item: DownloadItem
+    ) -> tuple[bool, DownloadUrlResponse | None]:
```

Comment:
Messy to return such a tuple. Maybe use enumeration instead of a bool to better indicate status?